File: actions_cpp/src/move_robot_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "robot_interfaces/action/move_robot.hpp"
#include "example_interfaces/msg/empty.hpp"

using MoveRobot = robot_interfaces::action::MoveRobot;
using MoveRobotGoalHandle = rclcpp_action::ClientGoalHandle<MoveRobot>;
using Empty = example_interfaces::msg::Empty;
using namespace std::placeholders;

class MoveRobotClientNode : public rclcpp::Node
{
public:
    MoveRobotClientNode() : Node("move_robot_client") 
    {
        move_robot_client_ = rclcpp_action::create_client<MoveRobot>(this, "move_robot");

        cancel_subscriber_ = this->create_subscription<Empty>("cancel_move", 10, std::bind(&MoveRobotClientNode::callback_cancel_move, this, _1));

        RCLCPP_INFO(this->get_logger(), "Action Client has been started");
    }
   
    void send_goal(int position, int velocity)
    {
        //waiting for the server
        move_robot_client_->wait_for_action_server();

        // define goal and velocity
        auto goal = MoveRobot::Goal();
        goal.position = position;
        goal.velocity = velocity;

        // ADD callbacks
        auto options = rclcpp_action::Client<MoveRobot>::SendGoalOptions();
        options.goal_response_callback = std::bind(&MoveRobotClientNode::goal_response_callback, this, _1);
        options.result_callback = std::bind(&MoveRobotClientNode::goal_result_callback, this, _1);
        options.feedback_callback = std::bind(&MoveRobotClientNode::goal_feedback_callback, this, _1, _2);
        
        //Send a goal 
        RCLCPP_INFO(this->get_logger(),"Sending a goal");
        move_robot_client_->async_send_goal(goal, options);
    }
    
private:

    // Callback to know if the goal was accpeted or rejected 
    void goal_response_callback(const MoveRobotGoalHandle::SharedPtr &goal_handle)
    {
        if (!goal_handle) {
            RCLCPP_INFO(this->get_logger(), "Goal got rejected");
        }
        else
        {
            this->goal_handle_ = goal_handle;
            RCLCPP_INFO(this->get_logger(),"Goal is accepted");
        }
    }

    //callback to receive the result once the goal is done 
    void goal_result_callback(const MoveRobotGoalHandle::WrappedResult &result)
    {
        auto status = result.code;
        if (status == rclcpp_action::ResultCode::SUCCEEDED){
            RCLCPP_INFO(this->get_logger(),"SUCCEEDED");
        }
        else if (status == rclcpp_action::ResultCode::CANCELED){
            RCLCPP_INFO(this->get_logger(),"CANCELLED");
        }
        else if (status == rclcpp_action::ResultCode::ABORTED)
        {
            RCLCPP_INFO(this->get_logger(),"ABORTED");
        }
        int position = result.result->position;
        std::string message = result.result->message;
        RCLCPP_INFO(this->get_logger(), "Position: %d", position);
        RCLCPP_INFO(this->get_logger(), "Message: %s", message.c_str());
    }

    // To receive feedback during goal execution
    void goal_feedback_callback(const MoveRobotGoalHandle::SharedPtr &goal_handle, const std::shared_ptr<const MoveRobot::Feedback> feedback)
    {
        (void)goal_handle;
        int position = feedback->current_position;
        RCLCPP_INFO(this->get_logger(),"Feedback Position: %d", position);
    }

    // Self callback cancel move for cancel_subscriber_
    void callback_cancel_move(const Empty::SharedPtr msg)
    {
        (void)msg;
        cancel_goal();
    }
    void cancel_goal()
    {
        if(this->goal_handle_){
            this->move_robot_client_->async_cancel_goal(goal_handle_);
            goal_handle_.reset();
        }
    }

    rclcpp_action::Client<MoveRobot>::SharedPtr move_robot_client_;
    rclcpp::Subscription<Empty>::SharedPtr cancel_subscriber_ ; 
    MoveRobotGoalHandle::SharedPtr goal_handle_;

};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MoveRobotClientNode>();
    node->send_goal(75, 1);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: actions_py/actions_py/move_robot_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus
from robot_interfaces.action import MoveRobot
from example_interfaces.msg import Empty

class MoveRobotClientNode(Node):
    def __init__(self):
        super().__init__("move_robot_client")
        self.goal_handle_ = None
        self.move_robot_client_ = ActionClient(self,MoveRobot,"move_robot")
        self.get_logger().info("Action Client has been started")
        self.cancel_subscriber = self.create_subscription(Empty, "cancel_move", self.callback_cancel_move, 10)

    def send_goal(self, position, velocity):
         
        # Waiting For the server
        self.move_robot_client_.wait_for_server()

        # create a goal
        goal = MoveRobot.Goal()
        goal.position = position
        goal.velocity = velocity
        
        # Send the goal
        self.get_logger().info("Sending the Goal with position: " + str(position) + " velocity: " + str(velocity))
        self.move_robot_client_.send_goal_async(goal, feedback_callback=self.goal_feedback_callback).add_done_callback(self.goal_response_callback)
    
    # Callback to show whether the goal was accepted or rejected 
    def goal_response_callback(self, future):
        self.goal_handle_: ClientGoalHandle = future.result()
        if self.goal_handle_.accepted:
            self.get_logger().info("GOAL ACCEPTED") 
            self.goal_handle_.get_result_async().add_done_callback(self.goal_result_callback)
        else:
             self.get_logger().info("GOAL REJECTED")

    # Callback group to show the Result to the client 
    def goal_result_callback(self, future):
        status = future.result().status
        result = future.result().result
        if status == GoalStatus.STATUS_SUCCEEDED:
             self.get_logger().info("SUCCESS")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error("ABORTED")  
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().error("CANCELLED")
        self.get_logger().info("Position: " + str(result.position))
        self.get_logger().info("Message: " + str(result.message))

    # Adding feedback for the client:
    def goal_feedback_callback(self, feedback_msg):
        position = feedback_msg.feedback.current_position
        self.get_logger().info("Feedback: " + str(position))

    def callback_cancel_move(self, msg):
        self.cancel_goal()

    def cancel_goal(self):
        if self.goal_handle_ is not None:
            self.get_logger().info("Send a Cancel request")
            self.goal_handle_.cancel_goal_async()
        
def main(args=None):
        rclpy.init(args=args)
        node = MoveRobotClientNode()
        node.send_goal(75, 1)
        rclpy.spin(node)
        rclpy.shutdown()   
if __name__ == "__main__": 
    main()